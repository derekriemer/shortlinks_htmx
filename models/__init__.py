import os
from typing import Annotated
from fastapi import Depends
from sqlmodel import create_engine, Session, SQLModel

sqlite_file_name = "golinks.sqlite"
sqlite_url = f"sqlite:///{sqlite_file_name}"
connect_args = {"check_same_thread": False}

engine = create_engine(sqlite_url, connect_args=connect_args, echo=True)


def create_db_and_tables():
    SQLModel.metadata.create_all(engine)


def get_session():
    try:
        with Session(engine) as session:
            yield session
    finally:
        session.close()


SessionDep = Annotated[Session, Depends(get_session)]
