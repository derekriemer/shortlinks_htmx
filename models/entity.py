from typing import Optional
from datetime import datetime
from sqlmodel import SQLModel, Field, Relationship


class Link(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    url: str

    shortcut: list["Shortcut"] = Relationship(back_populates="link")


class Shortcut(SQLModel, table=True):
    name: str = Field(primary_key=True)
    link_id: int = Field(foreign_key="link.id")
    created_at: Optional[datetime] = Field(default_factory=datetime.utcnow)
    access_count: int = Field(default=0)
    last_accessed: Optional[datetime] = None

    link: Link | None = Relationship(back_populates="shortcut")
