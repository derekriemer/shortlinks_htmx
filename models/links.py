from typing import Optional
from sqlmodel import SQLModel, Field, Relationship
from .shortcut import Shortcut


class Link(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    url: str

    shortcuts: list["Shortcut"] = Relationship(back_populates="link")
