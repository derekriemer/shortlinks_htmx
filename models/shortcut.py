from typing import Optional
from datetime import datetime
from sqlmodel import SQLModel, Field, Relationship
from .links import Link


class Shortcut(SQLModel, table=True):
    name: str = Field(primary_key=True)
    link_id: int = Field(foreign_key="link.id")
    created_at: Optional[datetime] = Field(default_factory=datetime.utcnow)
    access_count: int = Field(default=0)
    last_accessed: Optional[datetime] = None

    link: Optional[Link] = Relationship(back_populates="shortcuts")
