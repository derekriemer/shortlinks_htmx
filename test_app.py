from sqlmodel import Session, select, text
from fastapi.testclient import TestClient

from models.entity import Link, Shortcut

test_shortcuts = [
    {


        "name": "derek",
        "url": "https://derekriemer.com",
    },
    {
        "name": "example",
        "url": "https://example.com",
    }
]


def fill_db(session: Session):
    for shortcut in test_shortcuts:
        session.add(
            Shortcut(name=shortcut['name'], link=Link(url=shortcut['url'])))
    session.commit()


def test_get_shortcuts_table(session: Session, client: TestClient, verify):
    fill_db(session)
    response = client.get("/htmx/shortcuts")
    verify(response.text)
    assert response.status_code == 200

# for each of the create, update operations, we have two endpoints.
# an endpoint to retreive dataa form to do operation.
# An endpoint to receive data from the form.
# This because it's not really feasible to have fastapi handle both post and get in one handler.


def test_add_page(client: TestClient, verify):
    response = client.get("/htmx/add/", headers={
        "hx-request": "true",
    })
    verify(response.text)
    assert response.status_code == 200


def test_add_form_commits_data(client: TestClient, session: Session, verify):
    response = client.post("/htmx/add_form/", data={
        "name": test_shortcuts[0]['name'],
        "url": test_shortcuts[0]['url'],
    })
    assert response.status_code == 200
    verify(response.text)
    shortcut = session.exec(select(Shortcut)).one()
    assert shortcut.name == test_shortcuts[0]['name']
    assert shortcut.link.url == test_shortcuts[0]['url']


def test_edit_page(session: Session, client: TestClient, verify):
    fill_db(session)
    response = client.get("/htmx/edit/", params={"name": "derek"})
    assert response.status_code == 200
    verify(response.text)


def test_edit_form(session: Session, client: TestClient, verify):
    fill_db(session)
    response = client.post("/htmx/edit_form/", params={
        "old_name": "derek",
    }, data={
        "name": "me",
        "url": "https://derekriemer.com",
    })
    assert response.status_code == 200
    verify(response.text)
    shortcut = session.exec(select(Shortcut).where(
        Shortcut.name == "me")).one()
    assert shortcut.name == "me"
    assert shortcut.link.url == "https://derekriemer.com"


def test_delete_page(client: TestClient, session: Session):
    fill_db(session)
    response = client.get("/htmx/delete/", params={"name": "example"})
    assert response.status_code == 200
