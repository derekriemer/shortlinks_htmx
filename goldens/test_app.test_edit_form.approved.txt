<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8" />
  <title>
Home
</title>
  <script src="/assets/htmx.min.js"></script>
</head>

<body>

  <main hx-boost="true">
    
<label for="filter-input">
    Filter
</label>
<input type="text" name="filter" placeholder="Filter" id="filter-input" hx-get="/htmx/shortcuts"
    hx-trigger="input changed delay:500ms, keyup[key=='Enter']" hx-target="#link-table" hx-swap="outerHTML">

<div id="link-table">
    

    <table>
        <thead>
            <tr>
                <th>shortcut</th>
                <th>Url</th>
                <th>Actions</th>
            </tr>
        </thead>
        <tbody>
            
            <tr>
                <th>
                    example
                </th>
                <td>
                    https://example.com
                </td>
                <td>
                    <a href="/htmx/edit?name=example">
                        Edit
                    </a>
                    <a href="/htmx/delete?name=example"
                        hx-confirm=" Are you sure you want to delete this shortcut?" , hx-target="closest tr"
                        hx-swap="delete">
                        Delete
                    </a>
                </td>
            </tr>
            
            <tr>
                <th>
                    me
                </th>
                <td>
                    https://derekriemer.com
                </td>
                <td>
                    <a href="/htmx/edit?name=me">
                        Edit
                    </a>
                    <a href="/htmx/delete?name=me"
                        hx-confirm=" Are you sure you want to delete this shortcut?" , hx-target="closest tr"
                        hx-swap="delete">
                        Delete
                    </a>
                </td>
            </tr>
            
        </tbody>
    </table>
    
</div>

<button hx-get="/htmx/add" hx-target="body" hx-swap="outerHTML">
    Add Route

</button>
</main>
</body>

</html>
